File: Services/DnsService.cs
using System;
using System.Diagnostics;
using System.Threading.Tasks;
using InvolveX.Cli.Utils;

using System.Runtime.Versioning;

namespace InvolveX.Cli.Services
{
    [SupportedOSPlatform("windows")]
    public class DnsService
    {
        private readonly LogService _logService;

        public DnsService(LogService logService)
        {
            _logService = logService;
        }

        public async Task<bool> SetDns(string primaryDns, string secondaryDns)
        {
            if (!WindowsUtils.IsUserAnAdmin())
            {
                _logService.Log("Error: Administrator privileges are required to set DNS servers.");
                return false;
            }

            try
            {
                // Set primary DNS
                var result1 = await RunNetshCommand(
                    $"interface ip set dns name=\"Ethernet\" static {primaryDns} primary");
                if (result1.ExitCode != 0) return false;

                // Add secondary DNS
                var result2 = await RunNetshCommand(
                    $"interface ip add dns name=\"Ethernet\" {secondaryDns} index=2");
                return result2.ExitCode == 0;
            }
            catch (Exception ex)
            {
                _logService.Log($"Error setting DNS: {ex.Message}");
                return false;
            }
        }

        public async Task<bool> ResetDns()
        {
            if (!WindowsUtils.IsUserAnAdmin())
            {
                _logService.Log("Error: Administrator privileges are required to reset DNS servers.");
                return false;
            }

            try
            {
                var result = await RunNetshCommand("interface ip set dns name=\"Ethernet\" dhcp");
                return result.ExitCode == 0;
            }
            catch (Exception ex)
            {
                _logService.Log($"Error resetting DNS: {ex.Message}");
                return false;
            }
        }

        private async Task<ProcessResult> RunNetshCommand(string arguments)
        {
            var process = new Process
            {
                StartInfo = new ProcessStartInfo
                {
                    FileName = "netsh",
                    Arguments = arguments,
                    RedirectStandardOutput = true,
                    RedirectStandardError = true,
                    UseShellExecute = false,
                    CreateNoWindow = true
                }
            };

            process.Start();
            string stdout = await process.StandardOutput.ReadToEndAsync();
            string stderr = await process.StandardError.ReadToEndAsync();
            await process.WaitForExitAsync();

            _logService.Log($"[netsh STDOUT]: {stdout}");
            _logService.Log($"[netsh STDERR]: {stderr}");

            return new ProcessResult(process.ExitCode, stdout, stderr);
        }

        private class ProcessResult
        {
            public int ExitCode { get; } // Changed from private set to public get
            public string StdOut { get; } // Changed from private set to public get
            public string StdErr { get; } // Changed from private set to public get

            public ProcessResult(int exitCode, string stdOut, string stdErr)
            {
                ExitCode = exitCode;
                StdOut = stdOut;
                StdErr = stdErr;
            }
        }
    }
}

File: services/DnsService.js
const { spawn } = require('child_process');

class DnsService {
    constructor(logService) {
        this.logService = logService;
    }

    async setDns(primaryDns, secondaryDns = '') {
        try {
            this.logService.log(`Setting DNS to ${primaryDns}${secondaryDns ? ` and ${secondaryDns}` : ''}`);

            // Set primary DNS
            const primaryResult = await this.runNetshCommand(
                `interface ip set dns name="Ethernet" static ${primaryDns} primary`
            );

            if (primaryResult.code !== 0) {
                this.logService.log('Failed to set primary DNS');
                return false;
            }

            // Set secondary DNS if provided
            if (secondaryDns) {
                const secondaryResult = await this.runNetshCommand(
                    `interface ip add dns name="Ethernet" ${secondaryDns} index=2`
                );

                if (secondaryResult.code !== 0) {
                    this.logService.log('Failed to set secondary DNS');
                    return false;
                }
            }

            this.logService.log('DNS settings updated successfully');
            return true;
        } catch (error) {
            this.logService.log(`Error setting DNS: ${error.message}`);
            return false;
        }
    }

    async resetDns() {
        try {
            this.logService.log('Resetting DNS to DHCP');

            const result = await this.runNetshCommand('interface ip set dns name="Ethernet" dhcp');

            if (result.code === 0) {
                this.logService.log('DNS settings reset successfully');
                return true;
            } else {
                this.logService.log('Failed to reset DNS settings');
                return false;
            }
        } catch (error) {
            this.logService.log(`Error resetting DNS: ${error.message}`);
            return false;
        }
    }

    async runNetshCommand(args) {
        return new Promise((resolve, reject) => {
            const process = spawn('netsh', args.split(' '), {
                stdio: ['pipe', 'pipe', 'pipe'],
                shell: true
            });

            let stdout = '';
            let stderr = '';

            process.stdout.on('data', (data) => {
                stdout += data.toString();
            });

            process.stderr.on('data', (data) => {
                stderr += data.toString();
            });

            process.on('close', (code) => {
                this.logService.log(`[netsh STDOUT]: ${stdout}`);
                this.logService.log(`[netsh STDERR]: ${stderr}`);

                resolve({
                    code,
                    stdout,
                    stderr
                });
            });

            process.on('error', (error) => {
                reject(error);
            });
        });
    }
}

module.exports = DnsService;
